Added subscription assertions + docs (#2343)
nuwave/lighthouse: docs/master/testing/phpunit.md

    {
        parent::setUp();
+       $this->bootRefreshesSchemaCache();

> +       $this->setupSubscriptionEnvironment();

Reviewer: ```suggestion
+       $this->setupSubscriptionEnvironment(); // Only if you are using subscriptions
```

}
```

You can run a subscription query the same way. 

```php
public function testPostsSubscription(): void
{
    $response = $this->graphQL(/** @lang GraphQL */ '
    {

>         subscription OnPostCreated {

Reviewer: ```suggestion
        subscription {
```

}
```

### Subscription Assertions

#### Testing Authorization

> Once you do your subscription query and get a `$response`, you can run the following assertions on it.

Reviewer: ```suggestion
Once you do your subscription query and get a `TestResponse $response`, you can run the following assertions on it.
```

> If you want to make sure the current user is authorized to join a subscription you can use `assertGraphQLSubscriptionAuthorized`

Reviewer: ```suggestion
If you want to make sure the current user is authorized to join a subscription:
```

```php
$response->assertGraphQLSubscriptionAuthorized($this);
```

> If you want to make sure the current user is NOT authorized to join a subscription you can use `assertGraphQLSubscriptionNotAuthorized`

Reviewer: ```suggestion
If you want to make sure the current user is NOT authorized to join a subscription:
```

```php
$response->assertGraphQLSubscriptionNotAuthorized($this);
```

#### Testing Broadcast

Once you do your subscription query and get a `TestResponse $response`, you can run the following assertions on it.

> To assert a subscription actually received a broadcast:

Reviewer: ```suggestion
To assert the subscription actually received some broadcasts:
```

> Once you do your subscription query and get a `$response`, you can run the following assertions on it.

Reviewer: ```suggestion
Once you do your subscription query and get a `TestResponse $response`, you can run the following assertions on it.
```

> For asserting a broadcast you can use `assertGraphQLBroadcasted`

Reviewer: ```suggestion
To assert a subscription actually received a broadcast:
```

```

> // any other way to broadcast would laso work

Reviewer: ```suggestion
// any other way to broadcast would also work
```

Subscription::broadcast('postUpdated', ['title' => 'foo']);

> Subscription::broadcast('postUpdated', ['title' => 'bar']);

Reviewer: Can this be done after the call to the subscription has been made?

Author: Yes exactly, first something like this should be called 

```php
public function testPostsSubscription(): void
{
    $response = $this->graphQL(/** @lang GraphQL */ '
    {
        subscription OnPostCreated {
            onPostCreated {
                title
            }
        }
    }
    ');
}
```

Then `Subscription::broadcast` and then `assertGraphQLBroadcasted`

 
$response->assertGraphQLBroadcasted([
    ['title' => 'foo'],
    ['title' => 'bar'],
]);
```

> To assert a broadcast never happened:

Reviewer: ```suggestion
To assert the subscription received no broadcasts:
```

> For asserting that a broadcast never happened you can use `assertGraphQLNotBroadcasted`

Reviewer: ```suggestion
To assert a broadcast never happened:
```

```
// nothing that causes a broadcast to this channel
 
$response->assertGraphQLNotBroadcasted();
```

If you need more control over your broadcast assertion you can use `getGraphQLSubscriptionMock` which returns a [spy](http://docs.mockery.io/en/latest/reference/spies.html) and `getGraphQLSubscriptionChannelName`

```php

> $spy = $response->getGraphQLSubscriptionMock($this);

Reviewer: ```suggestion
$spy = $response->graphQLSubscriptionMock($this);
```

$spy->shouldNotHaveReceived('broadcast');
// or

> $spy->shouldNotHaveReceived('broadcast', function (Subscriber $subscriber, $broadcastedData) use ($response) {
>     $channel = $response->getGraphQLSubscriptionChannelName();
>     return $channel !== $subscriber->channel;
> });

Reviewer: ```suggestion
$spy->shouldNotHaveReceived('broadcast', function (Subscriber $subscriber, $broadcastedData) use ($response): bool {
    $channel = $response->getGraphQLSubscriptionChannelName();
    return $channel !== $subscriber->channel;
});
```
